### ISSUE TITLE
Do better at localization

### ISSUE THREAD

**Issue by micahflee** [micahflee/onionshare#669]
I think we should use a web-based localization framework such as Weblate or Transifex, so we can upload English-language locales and have people translate them into other languages, and others review their translations. We can then import the translations into OnionShare as a step in the release process.

Weblate is free software. and I know SecureDrop just set up [their own server](https://weblate.securedrop.club/). I'd rather not run our own servers of course.

Transifex is a centralized localization platform, and Tor Project uses it to localize Tor Browser, Tails, and other similar software. I think it might be a good idea to include OnionShare in Tor Project's account because they already have a dedicated army of translators that would be excellent to use as a resource. I'm at the Tor meeting now so I'll talk to people about this while I'm here.

It's possible that once this is implemented we'll have to refactor how OnionShare does translations to use a standard format that Weblate or Transifex supports.

**Comment by comradekingu** [comment]
Weblate is not only a great idea for being copyleft software, it is also a better tool, producing better results, that unlike Transifex, doesn't spy on its users.

https://hosted.weblate.org/ is a centralized hub for libre software projects that use Weblate, that are not self-hosted. Though sheer numbers is no indication of quality, it has more members than the Localization lab, which consequentially, you may be aware the Tor project is extending a hand to help in the process of moving away from Transifex.

https://docs.weblate.org/en/latest/devel/index.html is the relevant part of the documentation.

The quality of the current strings are very good, I had a go at trying to improve matters while checking #676

**Comment by mig5** [comment]
We've had another vote for Weblate in #730

**Comment by scootergrisen** [comment]
I dont care if its Transifex or Weblate. One of the other is fine.

**Comment by micahflee** [comment]
We're using Hosted Weblate now! @emmapeel2 has volunteered to manage the translations community.

https://hosted.weblate.org/projects/onionshare/

**Comment by scootergrisen** [comment]
Cool! :smile_cat: 
Lets me test it a bit...

**Comment by mig5** [comment]
Thanks @emmapeel2! :raised_hands:

**Comment by scootergrisen** [comment]
Component name is missing a "n" in "Traslations" but maybe call it something else.
Could change project name from "Onionshare" to "OnionShare".
If the strings from https://github.com/micahflee/onionshare/blob/develop/install/onionshare.desktop could be added to that would be fine so translations are in the same place.

    Comment=Share a file securely and anonymously over Tor
    Keywords=tor;anonymity;privacy;onion service;file sharing;file hosting;

Maybe add a new component called onionshare.desktop or something like that.

**Comment by comradekingu** [comment]
Can't do better than @emmapeel2 :) I tried my hand at improving the strings #751, and preliminarily went through the [wiki](https://github.com/micahflee/onionshare/wiki).

**Comment by scootergrisen** [comment]
Maybe we should protect the translations against actidential changes.
Like in another project one time i searched for some strings and changed them quickly but then i saw i have changed them in other languages by mistake because the Weblate search is a bit funny.
So for me i only translate to danish so maybe it makes sense that i can only change danish translation.

**Comment by emmapeel2** [comment]
Thanks for the support fellow translators, we meet again!
I will try to do my best to get OnionShare in more languages.

Scootergrissen, that looks like a bug from Weblate. Maybe you want to let them know? https://github.com/WeblateOrg/weblate/issues

**Comment by comradekingu** [comment]
Being able to change other languages is one of the great things with Weblate.
I regularly fix, among other things, all the Python string errors on the whole platform this way.
All admins of OnionShare on Weblate can change this behaviour, though IMO it is better to keep an eye out or react to it instead. @emmapeel2 If you make "kingu" an admin there I can upload screenshots from OnionShare to serve as contextual information for the strings.

**Comment by comradekingu** [comment]
@emmapeel2 The translated nb_NO component just got reset, and it has some translated strings, with a revision from 6 days ago, that isnt even the current translation.
https://hosted.weblate.org/translate/onionshare/translations/nb_NO/?type=translated&offset=1

Danish was also fully translated, and that is now at 55%.

Can't find anything other than https://github.com/micahflee/onionshare/commit/dbae5016895f1e29a15188aed4f21d0afb3350b4 and some talk about tracking a different branch on the mailinglist and changing the component name.

**Comment by mig5** [comment]
The branch was switched to `develop`, which is intended.

https://hosted.weblate.org/projects/onionshare/translations/#information

`develop` branch is very far ahead of `master` and contains many new strings. So it's also expected that the number of translations has reduced.

**Comment by comradekingu** [comment]
@mig5 Fine by me :) Can we consider #751 so work is applied where work is needed?

**Comment by micahflee** [comment]
Hey @comradekingu would you be interested in joining the OnionShare [mailing list](https://lists.riseup.net/www/info/onionshare-dev)? We've been discussing a localization strategy there and your input would be great.

And sorry about not spending time on #751 yet. We just finished merging some major features into develop which completely changes nearly all of the strings. I'll respond to the PR soon!

And btw "HS" stands for "Hidden Service", which is what Tor onion services used to be called before they were renamed "onion services". And now that I look for it, the `wait_for_hs` string is actually no longer even used in OnionShare...

**Comment by scootergrisen** [comment]
I brought danish translation up to 100 % the other day and read through all the strings and made a lot of changes.
All the changes now seems to have reverted back to the old translation.
Did you take a backup before changing the translations?

**Comment by comradekingu** [comment]
@scootergrisen OnionShare isn't yet ready for translation, just made available for translation.
I did the exact same thing, and then put my energy into trying to get the source strings workable. Please review #786 :)
I think we can have it done and shipped in the upcoming version if we work quickly once improvements land.

**Comment by micahflee** [comment]
@scootergrisen I'm so sorry about that! We had several months worth of major new features (including many major new strings) in the develop branch, but when we first started the Weblate account it was tracking the master branch. We finally got it to start tracking the develop branch instead, but you must have translated the master branch. Don't worry it won't happen again!

Now I've also merged #786, so the weblate is totally ready now to go in and start translating.

**Comment by scootergrisen** [comment]
I don't have permission to change the strings on Weblate.

**Comment by comradekingu** [comment]
Neither do I. "kingu".

**Comment by micahflee** [comment]
Hmm, when I login to weblate and look at the project settings, the access control is set to "Public" which is defined as "Publicly visible and translatable" (and since it's an open source project, I don't have permission to even change it). It's not set to Protected, which is defined as "Publicly visible, only translatable for chosen users". So you both _should_ be able to translate, I'm not sure what the problem is.

@emmapeel2 do you think you can troubleshoot and figure out what's going on?

**Comment by emmapeel2** [comment]
I will look onto this. Sorry about the loss!

**Comment by emmapeel2** [comment]
> If the strings from https://github.com/micahflee/onionshare/blob/develop/install/onionshare.desktop could be added to that would be fine so translations are in the same place.

Unfortunately, desktop files are not supported in Weblate.

**Comment by scootergrisen** [comment]
The strings could be in the json file or another file and then extracted.
Like PO files can have strings for .desktop files and then i think gettext tools extract the strings and some how create the .desktop file.
Its better to have the strings together because i oftent see .desktop files with missing translations and no instruction telling translations to rememeber to also translate the .desktop file.

**Comment by micahflee** [comment]
Regarding translating the .desktop file, I just opened an issue #793

**Comment by scootergrisen** [comment]
This does not work well yet...
I can't download the original JSON file and translate it locally and upload it. Because the file does not contain the source strings and the untranslated strings so even if its 100% translated locally, when i upload it its not 100% translated.
I suggest using po files.

**Comment by emmapeel2** [comment]
@scootergrisen yes this files are not easy to translate locally and I also prefer .po files. Sorry about that. One thing you can do is to download the English file and use it as a cheatsheet. Granted, is not as comfortable as using for example Poedit.

**Comment by emmapeel2** [comment]
@scootergrisen however, i am able to download the translation file as a .po file, from https://hosted.weblate.org/projects/onionshare/translations/es/ and selecting and option from Files menu under the title.

Can you try to do that too, and upload the updated .po file afterwards?

**Comment by scootergrisen** [comment]
I dont feel like maybe messing the translation up right now and having to do it over again now that i got it at 100 % translated again.

**Comment by emmapeel2** [comment]
sure. maybe for the future translations!

**Comment by heartsucker** [comment]
I think it would we worth squashing the commits in the translation branch down to a single commit pre-release and not merging the whole branch into `develop`. You can preserve the authors/contributors in the commit message, but it cuts down on noise when using `git log` and when examining branches. This is (roughly) what SecureDrop does.

**Comment by emmapeel2** [comment]
> I think it would we worth squashing the commits in the translation branch down to a single commit pre-release and not merging the whole branch into `develop`. 

Yes, that would be a neat and clean way to go. We always have weblate repo to see the history if we want, and we can honor attribution on the squash commit.
